src/client/sandbox/shadow-ui.js

```diff
@@ -155,7 +155,17 @@ export default class ShadowUI extends SandboxBase {
             if (typeof args[0] === 'string')
                 args[0] = NodeSandbox.processSelector(args[0]);
 
-            return ShadowUI._filterElement(nativeMethods.querySelector.apply(this, args));
+            var element = nativeMethods.querySelector.apply(this, args);
+
+            if (!element)
+                return null;
+
+            if (ShadowUI._filterElement(element))
+                return element;
+
+            var elements = ShadowUI._filterNodeList(nativeMethods.querySelectorAll.apply(this, arguments));
```

Comment:
- We don't need filter all nodes. We can check each node until first non-shadow ui element.
- Move code to separate method `this.overridenMethods.querySelector`